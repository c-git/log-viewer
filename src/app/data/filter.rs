use std::fmt::Display;

#[derive(Debug, Default, serde::Serialize, serde::Deserialize, PartialEq, Eq, Clone)]
#[serde(default)]
pub struct FilterConfig {
    pub search_key: String,
    pub filter_on: FilterOn,
    pub is_case_sensitive: bool,
    pub comparator: Comparator,
}

#[derive(Debug, Default, serde::Serialize, serde::Deserialize, PartialEq, Eq, Clone)]
pub enum FilterOn {
    #[default]
    Any,
    Field(FieldSpecifier),
}

#[derive(Debug, Default, serde::Serialize, serde::Deserialize, PartialEq, Eq, Clone)]
pub struct FieldSpecifier {
    pub name: String,
}

impl FilterOn {
    /// Returns `true` if the filter on is [`Any`].
    ///
    /// [`Any`]: FilterOn::Any
    #[must_use]
    pub fn is_any(&self) -> bool {
        matches!(self, Self::Any)
    }

    /// Returns `true` if the filter on is [`Field`].
    ///
    /// [`Field`]: FilterOn::Field
    #[must_use]
    pub fn is_field(&self) -> bool {
        matches!(self, Self::Field { .. })
    }
}

#[cfg_attr(test, derive(strum::EnumIter))]
#[derive(Debug, Default, serde::Serialize, serde::Deserialize, PartialEq, Eq, Clone, Copy)]
pub enum Comparator {
    LessThan,
    LessThanEqual,
    Equal,
    GreaterThan,
    GreaterThanEqual,
    NotEqual,
    #[default]
    Contains,
    NotContains,
}

impl Comparator {
    pub fn apply(&self, search_key: &str, value: &str) -> bool {
        match self {
            Comparator::LessThan => value < search_key,
            Comparator::LessThanEqual => value <= search_key,
            Comparator::Equal => value == search_key,
            Comparator::GreaterThan => value > search_key,
            Comparator::GreaterThanEqual => value >= search_key,
            Comparator::NotEqual => value != search_key,
            Comparator::Contains => value.contains(search_key),
            Comparator::NotContains => !value.contains(search_key),
        }
    }
}

impl Display for Comparator {
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        write!(
            f,
            "{}",
            match self {
                Comparator::LessThan => "Less Than",
                Comparator::LessThanEqual => "Less than equal",
                Comparator::Equal => "Equal",
                Comparator::GreaterThan => "Greater than",
                Comparator::GreaterThanEqual => "Greater than equal",
                Comparator::NotEqual => "Not equal",
                Comparator::Contains => "Contains",
                Comparator::NotContains => "Not contains",
            }
        )
    }
}

impl Display for FilterOn {
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        match self {
            FilterOn::Any => write!(f, "Any"),
            FilterOn::Field(name) => write!(f, "[Field Named: {name}]"),
        }
    }
}

impl Display for FieldSpecifier {
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        self.name.fmt(f)
    }
}
