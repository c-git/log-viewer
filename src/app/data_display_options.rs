use std::collections::{BTreeMap, BTreeSet};

use egui::Color32;

#[derive(serde::Deserialize, serde::Serialize, Debug, PartialEq, Eq)]
#[serde(default)] // if we add new fields, give them default values when deserializing old state
pub struct DataDisplayOptions {
    main_list_fields: Vec<String>,

    /// Lists fields to show last as they are not unique to a request
    common_fields: BTreeSet<String>,

    /// The field to use to highlight other related log entries
    ///
    /// WARNING: This must be a valid index into the list as this is assumed in method implementations
    emphasize_if_matching_field_idx: Option<usize>,

    /// Fields that should be colored based on their value. Key is field name
    pub colored_fields: BTreeMap<String, FieldColoringRules>,

    /// When set adds a field with this name and populates it with the row numbers (Skips record if field name already exists)
    pub row_idx_field_name: Option<String>,

    /// Controls how errors during file loading are treated
    pub row_parse_error_handling: RowParseErrorHandling,

    /// Used for optionally converting message levels to strings
    pub level_conversion: Option<LevelConversion>,

    /// Used for optionally including the size of messages
    pub row_size_config: Option<RowSizeConfig>,
}

#[derive(Default, serde::Deserialize, serde::Serialize, Debug, PartialEq, Eq)]
#[serde(default)]
pub struct FieldColoringRules {
    /// Matches a field value to color
    pub value_color_map: BTreeMap<String, Color32>,
}

#[derive(serde::Deserialize, serde::Serialize, Debug, PartialEq, Eq)]
pub enum RowParseErrorHandling {
    AbortOnAnyErrors,
    ConvertFailedLines {
        raw_line_field_name: String,
        /// If set the error message from the failure is placed in this field
        parse_error_field_name: Option<String>,
    },
}

#[derive(serde::Deserialize, serde::Serialize, Debug, PartialEq, Eq)]
#[serde(default)]
pub struct LevelConversion {
    /// Skips record if field name already exists
    pub display_field_name: String,
    /// Skips conversion if source field cannot be found
    pub source_field_name: String,
    pub convert_map: BTreeMap<i64, String>,
}

#[derive(Default, serde::Deserialize, serde::Serialize, Debug, PartialEq, Eq)]
#[serde(default)]
pub struct RowSizeConfig {
    pub field_name: String,
    pub units: SizeUnits,
}

#[derive(Default, serde::Deserialize, serde::Serialize, Debug, PartialEq, Eq)]
pub enum SizeUnits {
    Bytes,
    KB,
    MB,
    GB,
    TB,
    #[default]
    /// Is output as a String because it includes the unit and is the largest unit where the output value is >= 1
    Auto,
}
impl SizeUnits {
    pub(crate) fn convert(&self, row_size_in_bytes: usize) -> serde_json::Value {
        let scalar = match self {
            SizeUnits::Bytes => 1.0,
            SizeUnits::KB => 1024.0,
            SizeUnits::MB => 1024.0 * 1024.0,
            SizeUnits::GB => 1024.0 * 1024.0 * 1024.0,
            SizeUnits::TB => 1024.0 * 1024.0 * 1024.0 * 1024.0,
            SizeUnits::Auto => {
                // TODO 5: Rewrite with shifts for performance
                // Special handling see doc string for explanation
                let units = ["Bytes", "KB", "MB", "GB", "TB"];
                let mut last_index = 0;
                let mut scalar = 1.0f64;
                let row_size_in_bytes = row_size_in_bytes as f64;
                for i in 1..units.len() {
                    let new_scalar = scalar * 1024.0;
                    if (row_size_in_bytes / new_scalar) >= 1.0 {
                        last_index = i;
                        scalar = new_scalar;
                    } else {
                        // Last was as correct unit
                        break;
                    }
                }
                let result = row_size_in_bytes / scalar;
                return format!("{result:0>9.4} {}", units[last_index]).into();
            }
        };
        let result = row_size_in_bytes as f64 / scalar;
        result.into()
    }
}

impl DataDisplayOptions {
    pub fn main_list_fields(&self) -> &[String] {
        &self.main_list_fields
    }
    pub fn emphasize_if_matching_field_idx(&self) -> &Option<usize> {
        &self.emphasize_if_matching_field_idx
    }
    pub fn common_fields(&self) -> &BTreeSet<String> {
        &self.common_fields
    }
}

impl Default for DataDisplayOptions {
    fn default() -> Self {
        Self {
            // TODO 3: Add ability to show, select and reorder selected fields
            main_list_fields: [
                "row#",
                "level_str",
                "row_size",
                "time",
                "request_id",
                "otel.name",
                "msg",
            ]
            .into_iter()
            .map(String::from)
            .collect(),
            common_fields: [
                "elapsed_milliseconds",
                "file",
                "hostname",
                "http.flavor",
                "http.host",
                "http.method",
                "http.route",
                "http.scheme",
                "http.target",
                "http.user_agent",
                "level",
                "line",
                "name",
                "otel.kind",
                "pid",
                "req",
                "request_id",
                "res",
                "target",
                "time",
                "v",
            ]
            .into_iter()
            .map(String::from)
            .collect(),
            emphasize_if_matching_field_idx: Some(4),
            row_idx_field_name: Some("row#".to_string()),
            row_size_config: Some(RowSizeConfig {
                field_name: "row_size".to_string(),
                units: SizeUnits::Auto,
            }),
            row_parse_error_handling: Default::default(),
            level_conversion: Some(Default::default()),
            colored_fields: [(
                "level_str".to_string(),
                FieldColoringRules {
                    value_color_map: [
                        ("Trace".to_string(), Color32::from_rgb(150, 100, 200)),
                        ("Debug".to_string(), Color32::from_rgb(80, 140, 205)),
                        ("Info".to_string(), Color32::from_rgb(15, 175, 85)),
                        ("Warn".to_string(), Color32::from_rgb(210, 210, 20)),
                        ("Error".to_string(), Color32::from_rgb(220, 105, 105)),
                        ("Fatal".to_string(), Color32::from_rgb(255, 20, 20)),
                    ]
                    .into_iter()
                    .collect(),
                },
            )]
            .into_iter()
            .collect(),
        }
    }
}

impl Default for RowParseErrorHandling {
    fn default() -> Self {
        Self::ConvertFailedLines {
            raw_line_field_name: "msg".into(),
            parse_error_field_name: Some("parse_err".into()),
        }
    }
}

impl Default for LevelConversion {
    fn default() -> Self {
        // See bunyan levels https://github.com/trentm/node-bunyan?tab=readme-ov-file#levels and note rust only goes up to Error
        let convert_map = vec![
            (60, "Fatal".to_string()),
            (50, "Error".to_string()),
            (40, "Warn".to_string()),
            (30, "Info".to_string()),
            (20, "Debug".to_string()),
            (10, "Trace".to_string()),
        ]
        .into_iter()
        .collect();
        Self {
            display_field_name: "level_str".into(),
            source_field_name: "level".into(),
            convert_map,
        }
    }
}
