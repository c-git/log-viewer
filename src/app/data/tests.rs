use std::fmt::{Debug, Display};

use filter::Comparator;
use insta::glob;
use pretty_assertions::assert_eq;
use rstest::{fixture, rstest};
use strum::IntoEnumIterator;

use crate::app::data_display_options::DataDisplayOptions;

use super::*;

const PATH_PROJECT_ROOT: &str = "../../../";
const PATH_TEST_SAMPLES: &str = "tests/sample_logs/*.*";

/// Formats to test serializing with
///
/// Even though in the application only RON is used for serialization we do round
/// trip testing on Json because it helps identify problems that are format specific
/// and avoid unnecessary debugging
enum SerdeFormat {
    Ron,
    Json,
}

#[fixture]
pub(crate) fn insta_settings() -> insta::Settings {
    let mut result = insta::Settings::clone_current();
    let cwd = std::env::current_dir().expect("failed to get cwd");
    let path = cwd.join("tests").join("snapshots");
    result.set_snapshot_path(path);
    result
}

#[rstest]
fn deserialize_rows_from_string(insta_settings: insta::Settings) {
    glob!(PATH_PROJECT_ROOT, PATH_TEST_SAMPLES, |path| {
        let input = std::fs::read_to_string(path).unwrap();
        let data = Data::try_from(&input[..]).unwrap();
        let log_filename = path.file_name().unwrap().to_string_lossy().to_string();
        insta_settings.bind(|| insta::assert_ron_snapshot!(format!("{log_filename}_ron"), data));
        insta_settings.bind(|| insta::assert_yaml_snapshot!(format!("{log_filename}_yaml"), data));
        insta_settings
            .bind(|| insta::assert_debug_snapshot!(format!("{log_filename}_debug"), data));
    });
}

#[rstest]
#[case::ron(SerdeFormat::Ron)]
#[case::json(SerdeFormat::Json)]
fn round_trip_from_samples(#[case] serde_format: SerdeFormat) {
    // Function needed because rustfmt doesn't play nicely with formatting long strings in macros
    fn fail_with(path: impl Debug, row: usize, e: impl Debug, s: impl Display) -> LogRow {
        panic!(
            "failed to deserialize back into struct.\nFile: {path:?}\nRow: {row}\nError: {e:?}\nSerialized Data: {s}"
        )
    }

    glob!(PATH_PROJECT_ROOT, PATH_TEST_SAMPLES, |path| {
        let input = std::fs::read_to_string(path).unwrap();
        let rows_before = Data::try_from(&input[..]).unwrap();

        // Test individual rows
        for (i, row_before) in rows_before.rows_iter().enumerate() {
            let as_string = match serde_format {
                SerdeFormat::Ron => ron::to_string(&row_before).unwrap(),
                SerdeFormat::Json => serde_json::to_string(&row_before).unwrap(),
            };

            let row_after: LogRow = match serde_format {
                SerdeFormat::Ron => {
                    ron::from_str(&as_string).unwrap_or_else(|e| fail_with(path, i, e, as_string))
                }
                SerdeFormat::Json => serde_json::from_str(&as_string)
                    .unwrap_or_else(|e| fail_with(path, i, e, as_string)),
            };
            assert_eq!(&row_after, row_before);
        }

        // Test composition of all rows
        let as_string = match serde_format {
            SerdeFormat::Ron => ron::to_string(&rows_before).unwrap(),
            SerdeFormat::Json => serde_json::to_string(&rows_before).unwrap(),
        };
        let rows_after: Data = match serde_format {
            SerdeFormat::Ron => ron::from_str(&dbg!(as_string)).unwrap(),
            SerdeFormat::Json => serde_json::from_str(&dbg!(as_string)).unwrap(),
        };
        assert_eq!(rows_after, rows_before);
    });
}

pub fn create_log_row_no_extra() -> LogRow {
    let mut result = LogRow::default();
    result.data.insert("time".into(), "time value".into());
    result
        .data
        .insert("otel.name".into(), "HTTP GET /status".into());
    result
}

pub fn create_log_row_with_extra() -> LogRow {
    let mut result = create_log_row_no_extra();
    result.data.insert("http.status_code".into(), 200.into());
    result
}

#[rstest]
fn round_trip_from_manual(
    #[values(SerdeFormat::Ron, SerdeFormat::Json)] serde_format: SerdeFormat,
    #[values(create_log_row_no_extra(), create_log_row_with_extra())] before: LogRow,
) {
    let as_string = match serde_format {
        SerdeFormat::Ron => ron::to_string(&before).unwrap(),
        SerdeFormat::Json => serde_json::to_string(&before).unwrap(),
    };
    println!("Serialized data:\n{as_string}");
    let after: LogRow = match serde_format {
        SerdeFormat::Ron => ron::from_str(&as_string).unwrap(),
        SerdeFormat::Json => serde_json::from_str(&as_string).unwrap(),
    };
    assert_eq!(after, before);
}

#[rstest]
fn comparisons_specific_field(insta_settings: insta::Settings) {
    let row0 = create_log_row_no_extra();
    let row1 = create_log_row_with_extra();
    let mut data = Data {
        rows: vec![row0.clone(), row1.clone()],
        ..Default::default()
    };

    data.filter = Some(FilterConfig {
        search_key: "200".to_string(),
        filter_on: filter::FilterOn::Field(FieldSpecifier {
            name: "http.status_code".to_string(),
        }),
        is_case_sensitive: false,
        comparator: Default::default(),
    });

    let display_options = DataDisplayOptions::default();
    let common_fields = display_options.common_fields();

    for comparator in Comparator::iter() {
        data.filter.as_mut().unwrap().comparator = comparator;
        data.apply_filter(common_fields);
        insta_settings.bind(|| insta::assert_yaml_snapshot!(data));
    }
}

#[rstest]
fn comparisons_any(insta_settings: insta::Settings) {
    let row0 = create_log_row_no_extra();
    let row1 = create_log_row_with_extra();
    let mut data = Data {
        rows: vec![row0.clone(), row1.clone()],
        ..Default::default()
    };

    data.filter = Some(FilterConfig {
        search_key: "20".to_string(),
        filter_on: filter::FilterOn::Any,
        is_case_sensitive: false,
        comparator: Default::default(),
    });

    let display_options = DataDisplayOptions::default();
    let common_fields = display_options.common_fields();

    for comparator in Comparator::iter() {
        data.filter.as_mut().unwrap().comparator = comparator;
        data.apply_filter(common_fields);
        insta_settings.bind(|| insta::assert_yaml_snapshot!(data));
    }
}

#[test]
fn selected_maintenance_with_filtering() {
    let test_field = String::from("test field");
    let rows = (5..10)
        .map(|i| {
            let mut row = create_log_row_no_extra();
            row.data.insert(test_field.clone(), i.into());
            row
        })
        .collect();
    let mut data = Data {
        rows,
        ..Default::default()
    };
    let display_options = DataDisplayOptions::default();
    let common_fields = display_options.common_fields();

    // Set "7" as selected
    data.selected_row = Some(2);

    // Save selected row from before
    let expected = data
        .selected_row_data_as_slice(common_fields)
        .unwrap()
        .to_vec();

    data.filter = Some(FilterConfig {
        search_key: "7".to_string(),
        ..Default::default()
    });
    data.apply_filter(DataDisplayOptions::default().common_fields());

    // Test that 7 is still selected
    let actual = data
        .selected_row_data_as_slice(common_fields)
        .unwrap()
        .to_vec();

    assert_eq!(actual, expected);

    // Then reverse
    data.unfilter();

    // Test that 7 is still selected
    let actual = data
        .selected_row_data_as_slice(common_fields)
        .unwrap()
        .to_vec();

    assert_eq!(actual, expected);
}

#[test]
fn selected_unselected_when_not_present() {
    let test_field = String::from("test field");
    let rows = (5..10)
        .map(|i| {
            let mut row = create_log_row_no_extra();
            row.data.insert(test_field.clone(), i.into());
            row
        })
        .collect();
    let mut data = Data {
        rows,
        ..Default::default()
    };
    let display_options = DataDisplayOptions::default();
    let common_fields = display_options.common_fields();

    // Set "7" as selected
    data.selected_row = Some(2);

    // Filter for 6, so 7 is not included
    data.filter = Some(FilterConfig {
        search_key: "6".to_string(),
        ..Default::default()
    });
    data.apply_filter(DataDisplayOptions::default().common_fields());

    let actual = data.selected_row_data_as_slice(common_fields);

    assert!(actual.is_none());
}
