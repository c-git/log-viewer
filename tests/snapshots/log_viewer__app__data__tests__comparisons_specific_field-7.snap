---
source: src/app/data/tests.rs
expression: data
---
selected_row: ~
filter:
  search_key: "200"
  filter_on:
    Field:
      name: http.status_code
  is_case_sensitive: false
  comparator: Contains
rows:
  - data:
      otel.name: HTTP GET /status
      time: time value
  - data:
      http.status_code: 200
      otel.name: HTTP GET /status
      time: time value
filtered_rows:
  - 1
applied_filter:
  search_key: "200"
  filter_on:
    Field:
      name: http.status_code
  is_case_sensitive: false
  comparator: Contains
file_size: ""
